lib/BVHConstructionContext.js

```diff
@@ -0,0 +1,409 @@
+import * as THREE from '../node_modules/three/build/three.module.js';
+import { getLongestEdgeIndex } from './GeometryUtilities.js';
+
+const xyzFields = [ 'x', 'y', 'z' ];
+
+const getTriangleCount = ( geo ) => {
+
+	return geo.index ? ( geo.index.count / 3 ) : ( geo.attributes.position.count / 3 );
+
+};
+
+// TODO: This could probably be optimizied to not dig so deeply into an object
+// and reust some of the fetch values in some cases
+const getBufferGeometryVertexElem = ( geo, tri, vert, elem ) => {
+
+	return geo.attributes.position.array[ ( geo.index ? geo.index.array[ 3 * tri + vert ] : ( 3 * tri + vert ) ) * 3 + elem ];
+
+};
+
+// returns an array of size tris.length * 6 where triangle i maps to a
+// [x_min, x_max, y_min, y_max, z_min, z_max] tuple starting at index i * 6,
+// representing the minimum and maximum extent in each dimension of triangle i
+function computeBounds( geo ) {
+
+	const triCount = getTriangleCount ( geo );
+	const result = new Float32Array( triCount * 6 );
+
+	for ( let tri = 0; tri < triCount; tri ++ ) {
+
+		for ( let el = 0; el < 3; el ++ ) {
+
+			const a = getBufferGeometryVertexElem ( geo, tri, 0, el );
+			const b = getBufferGeometryVertexElem ( geo, tri, 1, el );
+			const c = getBufferGeometryVertexElem ( geo, tri, 2, el );
+			result[ tri * 6 + el * 2 ] = Math.min( a, b, c );
+			result[ tri * 6 + el * 2 + 1 ] = Math.max( a, b, c );
+
+		}
+
+	}
+
+	return result;
+
+}
+
+// returns an array of size tris.length * 3 where triangle i maps to an [x, y, z] triplet
+// starting at index i * 3, representing the centroid of triangle i
+function computeCentroids( geo ) {
+
+	const triCount = getTriangleCount ( geo );
+	const result = new Float32Array( triCount * 3 );
+
+	for ( let tri = 0; tri < triCount; tri ++ ) {
+
+		for ( let el = 0; el < 3; el ++ ) {
+
+			const a = getBufferGeometryVertexElem ( geo, tri, 0, el );
+			const b = getBufferGeometryVertexElem ( geo, tri, 1, el );
+			const c = getBufferGeometryVertexElem ( geo, tri, 2, el );
+			result[ tri * 3 + el ] = ( a + b + c ) / 3;
+
+		}
+
+	}
+
+	return result;
+
+}
+
+const SplitStrategy = {
+	get CENTER() {
+
+		return 0;
+
+	},
+	get AVERAGE() {
+
+		return 1;
+
+	},
+	get SAH() {
+
+		return 2;
+
+	}
+};
+
+const avgtemp = new THREE.Vector3();
+const centertemp = new THREE.Vector3();
+
+export default class BVHConstructionContext {
+
+	constructor( geo, options ) {
+
+		this.geo = geo;
+		this.options = options;
+		this.centroids = computeCentroids( geo );
+		this.bounds = computeBounds( geo );
+
+		// a list of every available triangle index
+		this.tris = new Array( getTriangleCount ( geo ) );
```

Comment:
What's the value of the `tris` array? I don't see it being changed anywhere so it seems to always hold the current index for that position in the array.

Reply:
Nevermind I'm seeing that `partition()` updates the `tris` array

Reply:
Next related question: Why not just operate directly on the `index` array rather than creating a new `tris` array and then writing the new order to the index buffer?

Reply:
We need each invocation of `createNode` to not only refer to the correct vertices, but also the correct precomputed bounding boxes and precomputed centroids. `tris` is an index into all three of those things at once. If we didn't have `tris`, each swap operation would have to not only swap three elements of the index (one per vertex), but six elements in the list of triangle bounding boxes and three elements in the list of centroids. I didn't measure it, but I suspected this was likely to be slower and was sure to be uglier than keeping this extra level of indirection.

Reply:
Yeah I see, now. This is more of an index into the index array.

Reply:
Yes, it's tricky to think about.